File: src/projectInstructions.ts
/**
 * Project-level instruction loading.
 *
 * Teams encode repo-local norms (build commands, do-not-touch paths, preferred
 * workflows, conventions) in a `DAINTREE.md` at the project root. We read it once
 * at startup and fold the content into the DYNAMIC prompt layer (message[1] via
 * {@link MainPromptContext.projectInstructions}) — never the cached base prefix —
 * so the prompt cache survives unchanged.
 *
 * The read is best-effort, mirroring the debug-log init philosophy: a missing or
 * unreadable file must NEVER block startup. Absence is the normal case for most
 * repos; an oversized or unreadable file returns a non-fatal warning the caller
 * surfaces however it surfaces other startup notices (render.warn / stderr).
 */
import path from "node:path";
import { stat, readFile } from "node:fs/promises";

/** The per-repo instruction file, read from the project root. */
export const PROJECT_INSTRUCTIONS_FILENAME = "DAINTREE.md";

/**
 * Hard cap on the instruction file. Large enough for rich team norms, small
 * enough that it can't blow the model's context or bias the cache. Files over
 * this size are skipped with a warning rather than truncated — silent truncation
 * would drop the team's instructions mid-sentence.
 */
export const PROJECT_INSTRUCTIONS_MAX_BYTES = 16 * 1024;

export interface ProjectInstructionsResult {
  /** The loaded instruction text (trimmed of surrounding whitespace). Absent when
   *  there's nothing to inject (no file, empty/whitespace-only, or skipped). */
  content?: string;
  /** A non-fatal warning to surface (oversized file, unreadable file). */
  warning?: string;
}

/**
 * Read `DAINTREE.md` from `projectPath`, if present. Never throws.
 *
 * Resolution and guards:
 *   - resolved against `projectPath` (the authoritative project root), not cwd
 *   - missing file / not a regular file → silently skip (the normal state)
 *   - larger than {@link PROJECT_INSTRUCTIONS_MAX_BYTES} → warn + skip
 *   - empty or whitespace-only → no content (no section gets rendered)
 *   - any other I/O error (e.g. EACCES) → warn + skip
 */
export async function loadProjectInstructions(
  projectPath: string,
): Promise<ProjectInstructionsResult> {
  const file = path.resolve(projectPath, PROJECT_INSTRUCTIONS_FILENAME);
  try {
    const info = await stat(file);
    if (!info.isFile()) return {};
    if (info.size > PROJECT_INSTRUCTIONS_MAX_BYTES) {
      return {
        warning: `Skipping ${PROJECT_INSTRUCTIONS_FILENAME}: ${info.size} bytes exceeds the ${PROJECT_INSTRUCTIONS_MAX_BYTES}-byte limit.`,
      };
    }
    const raw = await readFile(file, "utf8");
    // Re-check the actual byte length after reading: stat() and readFile() are two
    // syscalls, so a file that grew between them (or whose multibyte content the
    // stat size under-counts) must not slip past the cap. Belt-and-suspenders with
    // the stat() guard above.
    if (Buffer.byteLength(raw, "utf8") > PROJECT_INSTRUCTIONS_MAX_BYTES) {
      return {
        warning: `Skipping ${PROJECT_INSTRUCTIONS_FILENAME}: exceeds the ${PROJECT_INSTRUCTIONS_MAX_BYTES}-byte limit.`,
      };
    }
    const trimmed = raw.trim();
    if (trimmed.length === 0) return {};
    return { content: trimmed };
  } catch (err) {
    // ENOENT is the normal "no instructions for this repo" case — silent.
    if ((err as NodeJS.ErrnoException)?.code === "ENOENT") return {};
    const message = err instanceof Error ? err.message : String(err);
    return {
      warning: `Could not read ${PROJECT_INSTRUCTIONS_FILENAME}: ${message}`,
    };
  }
}

File: internal/projectinstructions/projectinstructions.go
// Package projectinstructions loads a repo's DAINTREE.md instruction file. It
// NEVER panics: a missing file is the normal case (silent skip); an oversized or
// unreadable file yields a warning and no content. Spec: docs/port/domain-config.md §3.
package projectinstructions

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

const (
	// Filename is the per-repo instruction file at the project root.
	Filename = "DAINTREE.md"
	// MaxBytes is the hard cap; oversized files are SKIPPED with a warning, not
	// truncated.
	MaxBytes = 16 * 1024
)

// Result carries trimmed instruction content and/or a warning. Both are empty
// when there is nothing to inject.
type Result struct {
	Content string
	Warning string
}

// Load reads <projectPath>/DAINTREE.md and returns a Result. It never returns an
// error — failures become warnings. The content is folded into the DYNAMIC prompt
// layer, never the cached base prefix.
func Load(projectPath string) Result {
	file := filepath.Join(projectPath, Filename)

	// Lstat (NOT Stat) so a SYMLINK is detected and rejected rather than followed: the
	// bound project is untrusted, and a repo could point DAINTREE.md at a secret file
	// (e.g. ~/.ssh/config) to exfiltrate its contents into the model prompt. We only ever
	// read a real regular file located at this exact path.
	info, err := os.Lstat(file)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return Result{} // normal case: no file
		}
		return Result{Warning: fmt.Sprintf("Could not read %s: %v", Filename, err)}
	}
	if info.Mode()&fs.ModeSymlink != 0 {
		return Result{Warning: fmt.Sprintf("Skipping %s: it is a symlink (not followed, to avoid reading a linked file into the prompt).", Filename)}
	}
	if !info.Mode().IsRegular() {
		return Result{} // silent skip (e.g. a directory named DAINTREE.md)
	}
	if info.Size() > MaxBytes {
		return Result{Warning: fmt.Sprintf("Skipping %s: %d bytes exceeds the %d-byte limit.", Filename, info.Size(), MaxBytes)}
	}

	raw, err := os.ReadFile(file)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return Result{}
		}
		return Result{Warning: fmt.Sprintf("Could not read %s: %v", Filename, err)}
	}
	// Re-check byte length: stat+read are two syscalls; guard against growth.
	if len(raw) > MaxBytes {
		return Result{Warning: fmt.Sprintf("Skipping %s: %d bytes exceeds the %d-byte limit.", Filename, len(raw), MaxBytes)}
	}

	trimmed := strings.TrimSpace(string(raw))
	if trimmed == "" {
		return Result{}
	}
	return Result{Content: trimmed}
}
